dozer/cogs/levels.py

```diff
@@ -0,0 +1,584 @@
+
+    async def update_server_settings_cache(self):
+        """Updates the server settings cache from the database"""
+        await asyncio.sleep(0.75)  # Give the database a few to update
+        self._guild_settings = {}
+        records = await GuildXPSettings.get_by()  # no filters, get all
+        for record in records:
+            self._guild_settings[record.guild_id] = record
+
+    async def update_level_role_cache(self):
+        """Updates level role cache from the database"""
+        await asyncio.sleep(0.75)  # Give the database a few to update
+        self._level_roles = {}
+        level_roles = await XPRole.get_by()
+        for role in level_roles:
+            if self._level_roles.get(role.guild_id):
+                self._level_roles[role.guild_id].append(role)
+            else:
+                self._level_roles[role.guild_id] = [role]
+
+    @staticmethod
+    def kwarg_parsing(message):
+        """Parse message into arguments based on a --"""
+        args = message.split(" ")
+        opts = {k.strip('-'): True if v.startswith('-') else v
+                for k, v in zip(args, args[1:] + ["--"]) if k.startswith('-')}
+        # a, b, c = (int(x) for x in (opts[y] for y in 'acb'))
+        return opts
+
+    @staticmethod
+    @functools.lru_cache(cache_size)
+    def total_xp_for_level(level):
+        """Compute the total XP required to reach the given level.
+        All members at this level have at least this much XP.
+        """
+        # https://github.com/Mee6/Mee6-documentation/blob/9d98a8fe8ab494fd85ec27750592fc9f8ef82472/docs/levels_xp.md
+        # > The formula to calculate how many xp you need for the next level is 5 * (lvl ^ 2) + 50 * lvl + 100 with
+        # > your current level as lvl
+        needed = 0
+        for lvl in range(level):
+            needed += 5 * lvl ** 2 + 50 * lvl + 100
+        return needed
+
+    @staticmethod
+    @functools.lru_cache(cache_size)
+    def level_for_total_xp(xp):
+        """Compute the level of a member with the given amount of total XP.
+        All members with this much XP are at or above this level.
+        """
+        # https://github.com/Mee6/Mee6-documentation/blob/9d98a8fe8ab494fd85ec27750592fc9f8ef82472/docs/levels_xp.md
+        # > The formula to calculate how many xp you need for the next level is 5 * (lvl ^ 2) + 50 * lvl + 100 with
+        # > your current level as lvl
+        lvl = 0
+        while xp >= 0:
+            xp -= 5 * lvl ** 2 + 50 * lvl + 100
+            lvl += 1
+        return lvl - 1
+
+    async def check_new_roles(self, guild, member, cached_member):
+        """Check and see if a member has qualified to get a new role"""
+        roles = self._level_roles.get(guild.id)
+        for level_role in roles:
+            role = guild.get_role(level_role.role_id)
+            if self.level_for_total_xp(cached_member.total_xp) >= level_role.level and role not in member.roles:
+                await member.add_roles(role)
```

Comment:
This makes one request per role, which will be ratelimited pretty heavily. This could collect the list of roles that need to be added in this loop and then add them all at once afterward.